t/pubsub/pubsub.t

```diff
@@ -0,0 +1,116 @@
+
+    if ((!defined $block->error_log) && (!defined $block->no_error_log)) {
+        $block->set_value("no_error_log", "[error]");
+    }
+
+    if (!defined $block->request) {
+        $block->set_value("request", "GET /t");
+    }
+});
+
+run_tests();
+
+__DATA__
+
+=== TEST 1: setup route by serverless
+--- config
+    location /t {
+        content_by_lua_block {
+            local data = {
+                {
+                    url = "/apisix/admin/routes/pubsub",
+                    data = {
+                        plugins = {
+                            ["serverless-pre-function"] = {
+                                phase = "access",
+                                functions =  {
+                                    [[return function(conf, ctx)
+                                        local core = require("apisix.core");
+                                        local pubsub, err = core.pubsub.new()
+                                        if not pubsub then
+                                            core.log.error("failed to initialize pub-sub module, err: ", err)
+                                            core.response.exit(400)
+                                            return
+                                        end
+                                        pubsub:on("cmd_kafka_list_offset", function (params)
+                                            return nil, "test"
+                                        end)
+                                        pubsub:wait()
+                                        ngx.exit(0)
+                                    end]],
+                                }
+                            }
+                        },
+                        uri = "/pubsub"
+                    },
+                },
+            }
+
+            local t = require("lib.test_admin").test
+
+            for _, data in ipairs(data) do
+                local code, body = t(data.url, ngx.HTTP_PUT, data.data)
+                ngx.say(code..body)
+            end
+        }
+    }
+--- response_body
+201passed
+
+
+
+=== TEST 2: hit route (with HTTP request)
+--- request
+GET /pubsub
+--- error_code: 400
+--- error_log
+failed to initialize pub-sub module, err: bad "upgrade" request header: nil
+
+
+
+=== TEST 3: connect websocket service
+--- config
+    location /t {
+        content_by_lua_block {
+            local lib_pubsub = require("lib.pubsub")
+            local test_pubsub = lib_pubsub.new_ws("ws://127.0.0.1:1984/pubsub")
+            local data = test_pubsub:send_recv_ws({
+                sequence = 0,
+                cmd_kafka_list_offset = {
+                    topic = "test",
+                    partition = 0,
+                    timestamp = -2,
+                },
+            })
+            if data and data.error_resp then
+                ngx.say("ret: ", data.error_resp.message)
+            end
+            test_pubsub:close_ws()
+        }
+    }
+--- response_body
+ret: test
```

Comment:
do we need more test cases to confirm the  basic `pub/sub` is working fine

Reply:
Tests have covered testable log points and possible error branches.